src/corelibs/U2View/src/ov_msa/exclude_list/MsaExcludeList.cpp

```diff
@@ -0,0 +1,593 @@
+    if (selectedItems.isEmpty()) {
+        sequenceView->setPlainText("");
+        return;
+    }
+    int excludeListRowId = getExcludeListRowId(selectedItems.first());
+    DNASequence sequence = sequenceByExcludeListRowId[excludeListRowId];
+    sequenceView->setPlainText(QString::fromUtf8(sequence.seq));
+}
+
+void MsaExcludeListWidget::showNameListContextMenu(const QPoint &) {
+    QMenu menu;
+    menu.addAction(moveToMsaAction);
+    menu.exec(QCursor::pos());
+}
+
+void MsaExcludeListWidget::moveMsaSelectionToExcludeList() {
+    const MaEditorSelection &selection = msaEditor->getSelection();
+    SAFE_POINT(!selection.isEmpty(), "Msa editor selection is empty!", );
+    QList<QRect> selectedRects = selection.getRectList();
+    QList<int> selectedMsaRowIndexes = msaEditor->getCollapseModel()->getMaRowIndexesFromSelectionRects(selectedRects);
+    QList<int> excludeListRowIds;
+    MultipleSequenceAlignmentObject *msaObject = msaEditor->getMaObject();
+    for (int msaRowIndex : qAsConst(selectedMsaRowIndexes)) {
+        excludeListRowIds << addMsaRowEntry(msaObject->getRow(msaRowIndex));
+    }
+
+    U2OpStatusImpl os;
+    U2UseCommonUserModStep userModStep(msaObject->getEntityRef(), os);
+    if (!os.hasError()) {
+        int versionBefore = msaObject->getObjectVersion();
+        msaObject->removeRows(selectedMsaRowIndexes);
+        // Exclude list re-uses msa row ids.
+        trackedUndoMsaVersions.insert(versionBefore, {true, excludeListRowIds});
+        trackedRedoMsaVersions.insert(msaObject->getObjectVersion(), {true, excludeListRowIds});
+
+        // Select the first row before the removed ones.
+        if (!msaEditor->isAlignmentEmpty()) {
+            int firstSelectedRowIndexBefore = selectedRects.first().top();
+            int newSelectedRowIndex = qMin(msaEditor->getCollapseModel()->getViewRowCount() - 1, firstSelectedRowIndexBefore);
+            msaEditor->selectRows(newSelectedRowIndex, 1);
+        }
+    }
+    updateState();
+}
+
+void MsaExcludeListWidget::moveExcludeListSelectionToMaObject() {
+    MultipleSequenceAlignmentObject *msaObject = msaEditor->getMaObject();
+    QList<DNASequence> sequences;
+    QList<int> excludeListRowIdsMovedToMsa;
+    QList<QListWidgetItem *> selectedItems = nameListView->selectedItems();
+    for (const QListWidgetItem *listItem : qAsConst(selectedItems)) {
+        sequences << getExcludeListRowSequence(listItem);
+        excludeListRowIdsMovedToMsa << getExcludeListRowId(listItem);
+    }
+    removeEntries(selectedItems);
+
+    int versionBefore = msaObject->getObjectVersion();
+    int insertionIndex = -1;  // Append by default.
+    if (!msaEditor->getSelection().isEmpty() && msaEditor->getRowOrderMode() == MaEditorRowOrderMode::Original) {
+        insertionIndex = msaEditor->getSelection().getRectList().last().bottom() + 1;
+    }
+    AddSequenceObjectsToAlignmentTask task(msaObject, sequences, insertionIndex, true);
```

Comment:
Tasks aren't supposed to be processed like this

Reply:
This is why I put TODO here. Will be addressed next in the separate PR to keep this more focused. Tests guards against the problem before it is fixed.

Repeating TODOs in the review have no value, isn't it?

Reply:
Fix for the TODO: https://github.com/ugeneunipro/ugene/pull/706 (quite a big update that worth to be tested separately)